Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+
+#### Example - PowerShell in server mode
+
+This example illustrates how PowerShell can be run in server mode.
+
+```powershell
+# Create PowerShell endpoint session (with default shell configuration)
+# Protocol handles messages through the child process stdIn/stdOut process pipes
+$proc = Start-Process -FilePath pwsh.exe -ArgumentList '-servermode -nologo -noprofile' -PassThru
+
+# Pass child process object to custom connection endpoint handler that reads/writes connection streams from stdIn/stdOut pipes
+[MyCustomConnection.CustomConnectionEndpoint]::SetEndPointProcess($proc)
+```
+
+### Implementation Example
+
+This is a simple example of how to create a custom ConnectionInfo and Transport implementation
+
+```csharp
+namespace MyCustomConnection
+{
+    // Derived custom connection information class
+    public sealed class CustomConnectionInfo : RunspaceConnectionInfo
+    {
+        // New multi-factor authentication switch for connection.
+        // Otherwise use existing 'ComputerName' and 'Credential' properties from derived class.
+        public SwitchParameter MFA { get; private set; }
+
+        // Message reader and writer.
+        internal StreamReader messageReader { get; private set; }
+        internal StreamWriter messageWriter { get; private set; }
+
+        // Override and implement this method that creates the custom client session transport with
+        // bound live connection data streams.
+        // This method is called from within the internal PowerShell remoting implementation.
+        public override BaseClientSessionTransportManager CreateClientSessionTransportManager(
+            Guid instanceId,
+            string sessionName,
+            PSRemotingCryptoHelper cryptoHelper)
+        {
+            // Establish connection with this connection information and create connection reader/writer
+            // data streams.
+            StartConnection();
+
+            return new CustomConnectionClientSessionTransportManager(
+                this,
+                instanceId,
+                cryptoHelper);
+        }
+    }
+
+    // Derived custom client session transport manager class.
+    public sealed class CustomConnectionClientSessionTransportManager : ClientSessionTransportManagerBase
+    {
+        CustomConnectionInfo _connectionInfo;
+
+        private CustomConnectionClientSessionTransportManager()
+        { }
+
+        // Constructor.
+        internal CustomConnectionClientSessionTransportManager(
+            CustomConnectionInfo connectionInfo,
+            Guid runspaceId,
+            PSRemotingCryptoHelper cryptoHelper) : base (runspaceId, cryptoHelper)
+        {
+            if (connectionInfo == null) { throw new PSArgumentException("connectionInfo"); }
+            _connectionInfo = connectionInfo;
+        }
+
+        // Start connection protocol.
+        // This is called by internal PowerShell remoting implementation.
+        public override void CreateAsync()
```

Comment:
This is fine for OutOfProc based transports where the communication happens over 2 pipes or a single bidirectional pipe but WSMan is a bit of an outlier. It has a few special messages like the `Create`, `Receive`, `Send`, `Command`, `Signal`, `Connect`, etc WSMan messages. Is the expectation of that transport to decode the OutOfProc fragment that PowerShell will write to the stream like `<Signal ../>` and convert it to the relevant WSMan one? Are things like `DisconnectAsync` and `ConnectAsync` expected to be in scope for the transport manager?

Reply:
That's what I feared but it definitely understandable. I've tried to reconcile the client end of a transport between WSMan and OutOfProc based clients and settled on something like this (written in Python) https://github.com/jborean93/pypsrp/blob/c6667a44a34a3d64c89e80ccae646f25c537b25a/psrp/connection_info.py#L138-L311. Still that has some differences with how the transport manager in C# is set out but it's just some food for thought.

While it might be difficult to integrate a 3rd party WSMan client using the proposed model (in/out streams) it may still be doable. I can potentially see the custom connection reading the streams itself and converting it to the WSMan envelopes required. From a brief investigation it can easily send the WSMan create body based on the first `Data` packet and craft the WSMan Command/Signal/Delete bodies based on the `Command`, `Signal`, `Close` packets respectively.

The downside is the lack of support for disconnected operations. While that's not working on Linux today it would have been nice for it to be implemented at some point in the future. Even just opening up that possibility would be nice but maybe it's enough for an MVP and disconnected operations could be exposed at a future point if desired.

Reply:
Connection disconnect/connect can still be added at some point.  But it is pretty complex (what do you do with streaming data?) and we would probably want a pretty good reason to do so.

Reply:
> what do you do with streaming data?

I would say it's the same with WSMan, there is no more streaming between the client and server until something reconnects and there are server side policies to determine what to do with the accumulated data during that disconnected phase. It would require the server end to support this as well which as least with WSMan is already there. Other transports obviously would require extra work (if at all possible).